File: python/leetcode/clone-graph.py
# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        if node is None:
            return None
        new_root = UndirectedGraphNode(node.label)
        cloned_map = {}
        cloned_map[new_root.label] = new_root
        q = []
        q.append(node)
        while (len(q) > 0):
            this_node = q.pop()
            for i in range(len(this_node.neighbors)):
                n = this_node.neighbors[i]
                if (n.label not in cloned_map):
                    cloned_map[n.label] = UndirectedGraphNode(n.label)
                    q.append(n)
                cloned_map[this_node.label].neighbors.append(cloned_map[n.label])
        return new_root

File: javascript/leetcode/clone-graph.js
/**
 * Definition for undirected graph.
 * function UndirectedGraphNode(label) {
 *     this.label = label;
 *     this.neighbors = [];   // Array of UndirectedGraphNode
 * }
 */

/**
 * @param {UndirectedGraphNode} graph
 * @return {UndirectedGraphNode}
 */
var cloneGraph = function(node) {
    if (node == null) { return null; }

    var newRoot = new UndirectedGraphNode(node.label);
    var map = {};
    map[newRoot.label] = newRoot;
    var q = [];
    q.push(node);

    while (q.length > 0) {
        var thisNode = q.pop();
        for (var i = 0; i < thisNode.neighbors.length; i++) {
            n = thisNode.neighbors[i];
            if (!map[n.label]) {
                map[n.label] = new UndirectedGraphNode(n.label);
                q.push(n);
            }
            map[thisNode.label].neighbors.push(map[n.label]);
        }
    }

    return newRoot;
};
